CASSANDRA-17510 streaming three min timeout

jonmeredith opened the pull request.

jonmeredith committed:
Fix streaming timeout for sessions that last longer than three minutes

Patch by Jon Meredith; reviewed by Caleb Rackliffe for CASSANDRA-17510

maedhroz commented on test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java:
```diff
@@ -34,8 +34,6 @@
-import org.apache.cassandra.net.AsyncStreamingInputPlus;
-import org.apache.cassandra.net.AsyncStreamingInputPlus.InputTimeoutException;
 
 import static org.apache.cassandra.utils.Clock.Global.nanoTime;
```
unused

maedhroz replied:
also `import java.util.concurrent.TimeUnit;` is unused

maedhroz commented on src/java/org/apache/cassandra/net/AsyncStreamingInputPlus.java:
```diff
@@ -217,23 +209,31 @@ public void close()
+        if (!channel.isRegistered())
+        {
+            throw new IllegalStateException("AsyncStreamingInputPlus did not complete close before channel was unregistered");
+        }
```
nit/style: I don't remember what our style guide says, but I'd be fine with getting rid of the braces for one-liners

jonmeredith committed:
Caleb review feedback

jonmeredith committed:
DNM: Setting CircleCI to high res config

jonmeredith committed:
java8-ify test

maedhroz commented on test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java:
```diff
@@ -161,6 +129,61 @@ public void available_ClosedButWithBytes()
+    public void rebufferAndCloseToleratesInterruption() throws InterruptedException
+    {
+        ByteBuf beforeInterrupt = channel.alloc().heapBuffer(1024);
+        beforeInterrupt.writeCharSequence("BEFORE", Charset.forName("us-ascii"));
```
```suggestion
        beforeInterrupt.writeCharSequence("BEFORE", StandardCharsets.US_ASCII);
```

jonmeredith replied:
Much better.

maedhroz commented on test/unit/org/apache/cassandra/net/AsyncStreamingInputPlusTest.java:
```diff
@@ -161,6 +129,61 @@ public void available_ClosedButWithBytes()
+            try
+            {
+                byte[] buffer = new byte[beforeInterruptBytes + afterInterruptBytes];
+                inputPlus.read(buffer, 0, beforeInterruptBytes + afterInterruptBytes);
```
Would it be worth it to throw a quick check for the number of bytes read?

```
int bytesRead = inputPlus.read(buffer, 0, beforeInterruptBytes + afterInterruptBytes);
assertEquals(buffer.length, bytesRead);
```

jonmeredith replied:
Can't hurt.

jonmeredith committed:
caleb review test feedback

jonmeredith committed:
review feedback from Benedict

decided to split isClosed int isProducerClosed and isConsumerClosed
to remove the volatile and simplify shutdown recovery logic.

jonmeredith committed:
clarified outdated javadoc

jonmeredith commented:
Merged
